src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
-						char* cl_pos=strrchr(mask,':');
-						// not perfect when line already contains wildcards, but works
-						if ((dot_pos-bs_pos>0) && (dot_pos-fs_pos>0) && (dot_pos-cl_pos>0))
-							strncat(mask, "*",DOS_PATHLENGTH - 1);
-						else strncat(mask, "*.*",DOS_PATHLENGTH - 1);
-					} else {
-					        safe_strcpy(mask, "*.*");
-				        }
-
-					RealPt save_dta=dos.dta();
-					dos.dta(dos.tables.tempdta);
-
-					bool res = DOS_FindFirst(mask, 0xffff & ~DOS_ATTR_VOLUME);
-					if (!res) {
-						dos.dta(save_dta);
-						break;	// TODO: beep
-					}
-
-					DOS_DTA dta(dos.dta());
-					char name[DOS_NAMELENGTH_ASCII];uint32_t sz;uint16_t date;uint16_t time;uint8_t att;
-
-					std::list<std::string> executable;
-					while (res) {
-						dta.GetResult(name,sz,date,time,att);
-						// add result to completion list
-
-						if (strcmp(name, ".") && strcmp(name, "..")) {
-							if (dir_only) { //Handle the dir only case different (line starts with cd)
-								if(att & DOS_ATTR_DIRECTORY) l_completion.push_back(name);
-							} else {
-							        if (is_executable_filename(name))
-								        // Prepend executables to a separate list
-								        // and place that list ahead of normal files.
-									executable.push_front(name);
-								else
-									l_completion.push_back(name);
-							}
-						}
-						res=DOS_FindNext();
-					}
-					/* Add executable list to front of completion list. */
-					std::copy(executable.begin(),executable.end(),std::front_inserter(l_completion));
-					it_completion = l_completion.begin();
-					dos.dta(save_dta);
+		}
+
+		case '\t': {
+			if (!completion.empty()) {
+				completion_index++;
+				if (completion_index == completion.size()) {
+					completion_index = 0;
 				}
+			} else {
+				completion = get_completions(command);
+				if (completion.empty()) {
+					break;
+				}
+				completion_index = 0;
 
-				if (l_completion.size() && it_completion->length()) {
-					for (;str_index > completion_index; str_index--) {
-						// removes all characters
-						move_cursor_back_one();
-						outc(' ');
-						move_cursor_back_one();
-					}
-
-					strcpy(&line[completion_index], it_completion->c_str());
-					len = (uint16_t)it_completion->length();
-					str_len = str_index = completion_index + len;
-					size = CMD_MAXLINE - str_index - 2;
-					DOS_WriteFile(STDOUT, (uint8_t *)it_completion->c_str(), &len);
+				auto delimiter = command.find_last_of(" \\");
+				if (delimiter == std::string::npos) {
+					completion_start = 0;
+				} else {
+					completion_start = delimiter + 1;
 				}
 			}
+
+			command.erase(completion_start);
+			command += completion[completion_index];
+			cursor_position         = command.size();
+			viewing_tab_completions = true;
 			break;
-		case 0x1b: /* Esc */
-			while (str_index < str_len) {
-				outc(line[str_index++]);
+		}
+
+		case '\b':
+			if (cursor_position > 0) {
+				command.erase(cursor_position - 1, 1);
```

Comment:
Do we expect that the cursor_position not exceed the command's size()? Suggest asserting this prior to the erase. (This helps document reasonable expectations for readers too).